[ONNX] Refactor op level debugging

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98509
* __->__ #97494

Fixes #97728 
Fixes #98622 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool
6. parametrized op_level_debug and dynamic_shapes into FX tests

titaiwangms committed:
[ONNX] Refactor op level debugging

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. 

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. trace_only onnxscript function is not supported due to lack of param_schema.
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. 

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. trace_only onnxscript function is not supported due to lack of param_schema.
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. 

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. trace_only onnxscript function is not supported due to lack of param_schema.
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.

[ghstack-poisoned]

justinchuby commented on torch/onnx/_internal/onnx_proto_utils.py:
```diff
@@ -301,14 +301,13 @@ def _convert_tensor_to_numpy(input: Any) -> Any:
     if isinstance(input, torch.dtype):
         return int(_type_utils.JitScalarType.from_dtype(input).onnx_type())
-    if isinstance(input, (tuple, list)):
+    if isinstance(input, Sequence):
```
Can be string?

BowenBao commented on torch/onnx/_internal/fx/exporter.py:
```diff
@@ -35,7 +35,14 @@ def _export(
+    static_reference_graph_module = passes.ShapeInferenceWithFakeTensor(
+        static_reference_graph_module
+    ).run(*args)
+    export_options.update(static_reference_graph=static_reference_graph_module.graph)
```
Should not stash arbitrary things under `export_options`. With new exporter api this should presumably live under `DynamoExporter`.

titaiwangms replied:
Sure, I can change this to a passing argument

titaiwangms replied:
Done

titaiwangms commented:
Looks like FakeTensorProp has behavior change aster https://github.com/pytorch/pytorch/pull/97204. Now that sym_size shape can't be propagated. We might need an alternative way if it's intended.

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. trace_only onnxscript function is not supported due to lack of param_schema.
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.
3. sym_size and built-in ops are not supported.

[ghstack-poisoned]

titaiwangms commented:
> Looks like FakeTensorProp has behavior change after #97204. Now that sym_size shape can't be propagated. We might need an alternative way if it's intended.

Added our own pass for FakeTensorProp which supports int/float/bool

titaiwangms commented on torch/onnx/_internal/fx/passes/shape_inference.py:
```diff
@@ -45,10 +47,63 @@ def to_fake_tensor(x):
+
+
+@compatibility(is_backward_compatible=False)
+class ValidationFakeProp(torch.fx.Interpreter):
```
@BowenBao @wschin . This is the solution for us, but I am not sure what stage we are with FakeTensorProp now. It seems it's been actively developed...

titaiwangms commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -62,7 +65,8 @@ def _run_test_with_fx_to_onnx_exporter_and_onnx_runtime(
         opset_version=opset_version,
         use_binary_format=True,
-        enable_dynamic_axes=True,
+        enable_dynamic_axes=False,
```
Disabling dynamic_axes in runtime to avoid SegFault.

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -77,6 +117,45 @@ def forward(self, x):
+            diagnostics.rules.fx_node_to_onnx,
+            diagnostics.levels.WARNING,
+            expected_error_node="aten.embedding.default",
+            expected_error_message="IndexError: index out of range in self",
```
@BowenBao A naive assertion from SARIF log. We can have an offline sync if you feel this should be a standalone rule, but if you think it's enough for now, it's fine with me.

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool

[ghstack-poisoned]

justinchuby commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -17,6 +20,7 @@
 import parameterized
 import torch
 import transformers  # type: ignore[import]
+from pytorch_test_common import skipDynamicFXTest
```
Import modules only

titaiwangms replied:
Done

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool

[ghstack-poisoned]

titaiwangms commented on test/onnx/pytorch_test_common.py:
```diff
@@ -156,6 +157,26 @@ def wrapper(self, *args, **kwargs):
     return skip_dec
 
 
+def skipDynamicFXTest(reason: str = ""):
```
Done. Would like to mention that this is the first skip_test that we use snake case though.

justinchuby replied:
that's ok

titaiwangms commented:
@BowenBao PTAL

BowenBao commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -29,15 +29,23 @@ def export_fx_to_onnx(
             enable_dynamic_axes=self.options.dynamic_shapes,
         ).run(*fx_module_args)
-
+        # Run FakeTensorProp to get fixed shape of nodes for op_level_debug purposes.
```
```suggestion
        # Run ShapeInferenceWithFakeTensor to get static shape of nodes for op_level_debug purposes.
```

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -0,0 +1,224 @@
+    torch_args: tuple,
+    torch_kwargs: dict,
+):
+    """Validate the op between ONNX Runtime and PyTorch."""
```
It helps to have a few words explaining what to expect would happen.

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -0,0 +1,224 @@
+        try:
+            ort_outputs = symbolic_fn(*input_onnx, **kwargs_onnx)
+        except ValueError as value_error:
+            # FIXME(titaiwang): This is caused by wronly split args/kwargs.
```
```suggestion
            # FIXME(titaiwang): This is caused by wrongly split args/kwargs.
```

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/op_validation.py:
```diff
@@ -0,0 +1,224 @@
+            flattened_torch_outputs, flattened_function_outputs
+        ):
+            try:
+                if not isinstance(function_output, np.ndarray):
```
nit: instead check that `function_output` is "onnxscript tensor" type?

titaiwangms replied:
In fact, this should be deleted.

titaiwangms replied:
Done

BowenBao commented on torch/onnx/_internal/fx/fx_exporter.py:
```diff
@@ -29,15 +29,23 @@ def export_fx_to_onnx(
+        static_reference_graph_module = torch.fx.Transformer(
+            decomposed_module
+        ).transform()
+        static_reference_graph_module = passes.ShapeInferenceWithFakeTensor(
```
Ideally I would like to merge these operations into one pass. That pass runs these steps internally. And what's returned is still the `decomposed_module`, but with extra information on static tensor shapes. We'd need a definition for exporter metadata.

That way the implementation details are hidden, and you won't need to pass two graph modules to `export_fx_to_onnxscript`, and assuming the two graphs are in sync.

I'm fine if we are not there yet. It might be a little back and forth, but for the current state I feel it might be better to move these code inside `export_fx_to_onnxscript`. This pass is only needed when `op_level_debug=True`. Bringing it local makes the "in sync assumption" part more natural.

titaiwangms replied:
So I merged the Transformer into ShapeInference pass. And embed the propagate `node_with_static_shape` into original node.meta["node_with_static_shape"]. We can work on the naming of this metadata, but I think this is the best solution we can have now. WDYT

titaiwangms replied:
The downside is that we still have that sync assumption of two graphs. I tried to FakeTensorProp value into node.meta["real_shape"] instead of node.meta["val"], and it doesn't work. I guess if we really want same module in same module out on this pass, we might have to write an Interpreter.

BowenBao replied:
> tried to FakeTensorProp value into node.meta["real_shape"] instead of node.meta["val"], and it doesn't work.

Feel free to call

BowenBao reviewed (commented):
Thanks for refactoring! A few comments for consideration.

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool
6. parametrized op_level_debug and dynamic_shapes into FX tests

[ghstack-poisoned]

titaiwangms commented on test/onnx/pytorch_test_common.py:
```diff
@@ -156,6 +157,29 @@ def wrapper(self, *args, **kwargs):
+    """Skip dynamic exporting test.
+
+    Args:
+        reason: The reason for skipping scripting test.
```
Done

titaiwangms commented on test/onnx/test_fx_to_onnx.py:
```diff
@@ -6,14 +6,57 @@
+    expected_error_node: str,
+    expected_error_message: str,
+):
+    rule_level_pairs = (rule.id, level.name.lower())
```
It's a bug. I added another condition into if

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -252,65 +249,10 @@ def _fill_tensor_meta(
-    complete_args: List[_type_utils.Argument] = []
-    complete_kwargs: Dict[str, _type_utils.Argument] = {}
+) -> Tuple[List[_type_utils.Argument], Dict[str, _type_utils.Argument]]:
+    """Separate input attributes from arguments."""
```
Hmm... Changed to another name

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -332,13 +279,28 @@ def _wrap_fx_args_as_onnxscript_args(
+    onnxscript_kwargs = filter_incompatible_and_dtype_convert_kwargs(complete_kwargs)
 
-    return (onnxscript_args, onnxscript_kwargs, complete_args, complete_kwargs)
+    return (onnxscript_args, onnxscript_kwargs)
```
Done

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -462,12 +425,10 @@ def _export_fx_node_to_onnxscript(
-            complete_kwargs,
-        ) = _wrap_fx_args_as_onnxscript_args(node, fx_name_to_onnxscript_value, tracer)
+        complete_args, complete_kwargs = _separate_input_attributes_from_arguments(node)
+        (onnx_args, onnx_kwargs) = _wrap_fx_args_as_onnxscript_args(
```
Done

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -476,21 +437,30 @@ def _export_fx_node_to_onnxscript(
+            ) = _separate_input_attributes_from_arguments(node_with_fixed_shape)
+            # TODO(titaiwang): Why some ops failed on op-level validation, but
+            # successfully exported?
+            (torch_args, torch_kwargs) = op_validation.wrap_fx_args_as_torch_args(
```
Done

titaiwangms commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -476,21 +437,30 @@ def _export_fx_node_to_onnxscript(
-                complete_args, complete_kwargs
+        if (
+            options.op_level_debug
+            and not isinstance(node.target, torch._ops.OpOverloadPacket)
```
Done

titaiwangms commented on torch/onnx/_internal/fx/passes/shape_inference.py:
```diff
@@ -45,10 +47,63 @@ def to_fake_tensor(x):
         return module
+
+
+@compatibility(is_backward_compatible=False)
```
Removed. seems fine.

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool
6. parametrized op_level_debug and dynamic_shapes into FX tests

[ghstack-poisoned]

titaiwangms commented:
Had CI fail on:
`worker 'gw1' crashed while running 'test/onnx/test_fx_to_onnx_with_onnxruntime.py::TestFxToOnnxWithOnnxRuntime_op_level_debug_False_dynamic_shapes_True::test_flatten_dynamic_axes'`

Not sure it's on the machine or the test...

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool
6. parametrized op_level_debug and dynamic_shapes into FX tests

[ghstack-poisoned]

justinchuby commented:
Most likely because of ORT. This may be useful: https://github.com/microsoft/onnx-script/pull/602

BowenBao commented on torch/onnx/_internal/fx/passes/fx_to_onnxscript.py:
```diff
@@ -476,21 +436,35 @@ def _export_fx_node_to_onnxscript(
+                node_with_fixed_shape_args,
+                node_with_fixed_shape_kwargs,
+            ) = _fill_in_default_kwargs(node_with_static_shape)
+            # TODO(titaiwang): Why some ops failed on op-level validation, but
```
gap between eager vs export-and-execute?

titaiwangms replied:
Legacy. I will remove it..

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -193,23 +262,163 @@ def __init__(self):
+            additional_test_inputs=[(y,)],
+        )
 
     def test_mutation(self):
```
Please do me a favor and skip this one for dynamic too #98622  🤣

titaiwangms replied:
Sure!

titaiwangms replied:
Hmmm The one reported is actually not dynamic. I will skip the whole test in this PR, and put skip_ORT_version on it in the next PR.

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes #98622 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool
6. parametrized op_level_debug and dynamic_shapes into FX tests

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Refactor op level debugging"


Fixes #97728 
Fixes #98622 
Fixes https://github.com/microsoft/onnx-script/issues/393

Provide op_level_debug in exporter which creates randomnied torch.Tensor based on FakeTensorProp real shape as inputs of both torch ops and ONNX symbolic function. The PR leverages on Transformer class to create a new fx.Graph, but shares the same Module with the original one to save memory.

The test is different from [op_correctness_test.py](https://github.com/microsoft/onnx-script/blob/main/onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py) as op_level_debug generating real tensors based on the fake tensors in the model.

Limitation:
1. Some of the trace_only function is not supported due to lack of param_schema which leads to arg/kwargs wronly split and ndarray wrapping. (WARNINGS in SARIF)
2. The ops with dim/indices (INT64) is not supported that they need the information(shape) from other input args.  (WARNINGS in SARIF)
3. sym_size and built-in ops are not supported. 
4. op_level_debug only labels results in SARIF. It doesn't stop exporter.
5. Introduce ONNX owning FakeTensorProp supports int/float/bool
6. parametrized op_level_debug and dynamic_shapes into FX tests

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge

BowenBao commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -193,34 +263,285 @@ def __init__(self):
+        " typing.Sequence[int], torch.Tensor], as [None, None]:"
+    )
+    def test_shufflenet_v2_dynamic_axes(self):
+        model = torchvision.models.shufflenet_v2_x0_5(pretrained=False)
```
How does this pass CI ... Seems torchvision is not imported from anywhere?

titaiwangms replied:
Isn't this skipped?

titaiwangms replied:
🫢

BowenBao replied:
no worries, I'm adding a skip decorator for torchvision if uninstalled in the other PR
